### PR TITLE
Speedup migration rendering

### PR DESCRIPTION
**Pull request by MarkusH**
https://code.djangoproject.com/ticket/24554

### CODE REVIEW

**Comment by claudep** [comment]
By just reading the code, this looks like nice optimization. Do you have any figure to demonstrate improvements?

**Comment by MarkusH** [comment]
No, unfortunately not yet. Working on a set of migrations that include "every" case.

**Comment by MarkusH** [comment]
@claudep I pushed a benchmark repo to [MarkusH/django-migrations-benchmark](https://github.com/MarkusH/django-migrations-benchmark). (Py3 only -- no `__init__.py` in the app directories).

This change makes a difference when a lot of migrations are already applied and only a few (50:50) need to be applied.

File: django/db/migrations/state.py [hunk 1]
```diff
@@ -35,6 +35,12 @@ def get_related_models_recursive(model):
     """
     Returns all models that have a direct or indirect relationship
     to the given model.
+
+    Relationships are either defined by explicit relational fields, like
+    ForeignKey, ManyToManyField or OneToOneField, or by inheriting from another
+    model where subclass are included but superclasses not. Note, however, that
+    a model inheriting from a concrete model also is related to its superclass
```
Comment (timgraham) [inline django/db/migrations/state.py]: also is -> is also

File: django/db/migrations/state.py [hunk 2]
```diff
@@ -35,6 +35,12 @@ def get_related_models_recursive(model):
     """
     Returns all models that have a direct or indirect relationship
     to the given model.
+
+    Relationships are either defined by explicit relational fields, like
+    ForeignKey, ManyToManyField or OneToOneField, or by inheriting from another
+    model where subclass are included but superclasses not. Note, however, that
```
Comment (timgraham) [inline django/db/migrations/state.py]: "subclasses are included"?

**Comment by timgraham** [comment]
Waiting on anything else to merge this?

**Comment by MarkusH** [comment]
I'll have a look and merge if I'm happy with it.

**Comment by MarkusH** [comment]
check docs

**Pull request merged**